### PR TITLE
Enhancing Multi-speed Coil Speed Level Control Actuators

### PR DESCRIPTION
**Pull request by xuanluo113**
Pull request overview
---------------------
Previously in the 9.6.0 release, we added the EMS actuators to control the speed level of the cooling and heating multispeed coils, namely `Coil:Cooling:DX:MultiSpeed` and `Coil:Heating:DX:Multispeed`. As an enhancement, we propose to extend the EMS functionality to a more general coil type in EnergyPlus, `Coil:Cooling:DX`, which will replace the single speed, two speed, variable speed, multi-speed, and multi-stage coils going forward.  

Result of the `testfiles/EMSLoadBasedMultiSpeedDXCoilOverrideControl.idf`

Screen Shot 2021-11-23 at 10 00 21 AM

Result of the `testfiles/EMSSetpointBasedMultiSpeedDXCoilOverrideControl.idf`

Screen Shot 2021-11-23 at 10 00 37 AM

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by mjwitte** [comment]
@xuanluo113 This is straightforward enough. If I'm following correctly, since the actuator is applied to the unitary system, there is no change in inputs for this, just doc and code changes to extend this to work with Coil:Cooling:DX. When updating the documentation, some reformatting would be helpful (the Unitary Equipment actuator section is a bit hard to navigate).

**Comment by xuanluo113** [comment]
The branch is ready for review with a new unit test and three test files. Results are illustrated in the overview.

**Comment by xuanluo113** [comment]
@mjwitte @Myoldmopar Mike and Edwin, this branch is ready for review, and I wonder if you could take a look. As for the documentation, I'm not sure what is a better place to add the EMS description other than the `Unitary Equipment`, since the actuators are based on unitary systems.

File: src/EnergyPlus/UnitarySystem.cc [hunk 1]
```diff
@@ -11933,7 +11934,64 @@ namespace UnitarySystems {
...
+                    (CoilType_Num == DataHVACGlobals::CoilDX_MultiSpeedCooling || CoilType_Num == DataHVACGlobals::CoilDX_Cooling)) {
+                    this->m_CoolingSpeedNum = ceil(this->m_EMSOverrideCoilSpeedNumValue);
+                    this->m_SpeedNum = this->m_CoolingSpeedNum;
+                    bool useMaxedSpeed = false;
+                    if (this->m_SpeedNum > this->m_NumOfSpeedCooling) {
+                        this->m_CoolingSpeedNum = this->m_NumOfSpeedCooling;
+                        this->m_SpeedNum = this->m_NumOfSpeedCooling;
+                        CycRatio = this->m_CoolingCycRatio = 1.0;
+                        PartLoadFrac = SpeedRatio = this->m_CoolingSpeedRatio = 1.0;
+                        this->m_CoilSpeedErrIdx++;
```
Comment (mjwitte) [inline src/EnergyPlus/UnitarySystem.cc:11947]: This index should not be incremented here. This is an index to this particular recurring error message. If it's zero `ShowRecurringWarningErrorAtEnd` will set it to the next available index. There should also be a regular warning message issued the first time this occurs (i.e. when `m_CoilSpeedErrIdx=0`) . See UnitarySystem.cc lines 9347ff for an example of how to use `ShowRecurringWarningErrorAtEnd`.

File: src/EnergyPlus/UnitarySystem.cc [hunk 2]
```diff
@@ -11933,7 +11934,64 @@ namespace UnitarySystems {
...
+                            this->m_CoilSpeedErrIdx,
+                            this->m_EMSOverrideCoilSpeedNumValue,
+                            this->m_EMSOverrideCoilSpeedNumValue,
+                            _,
+                            "",
+                            "");
+                    } else if (this->m_CoolingSpeedNum == 1) {
+                        SpeedRatio = this->m_CoolingSpeedRatio = 0.0;
+                        CycRatio = this->m_CoolingCycRatio = this->m_EMSOverrideCoilSpeedNumValue - floor(this->m_EMSOverrideCoilSpeedNumValue);
+                        if (useMaxedSpeed || this->m_CoolingCycRatio == 0) {
```
Comment (mjwitte) [inline src/EnergyPlus/UnitarySystem.cc]: Not sure this should be an `else if`. Looks like `useMaxedSpeed` will always be false here.

File: src/EnergyPlus/UnitarySystem.cc [hunk 3]
```diff
@@ -11933,7 +11934,64 @@ namespace UnitarySystems {
...
+                            this->m_CoolingCycRatio = 1;
+                        }
+                        PartLoadFrac = this->m_CoolingCycRatio;
+                    } else {
+                        CycRatio = this->m_CoolingCycRatio = 1.0;
+                        SpeedRatio = this->m_CoolingSpeedRatio = this->m_EMSOverrideCoilSpeedNumValue - floor(this->m_EMSOverrideCoilSpeedNumValue);
```
Comment (mjwitte) [inline src/EnergyPlus/UnitarySystem.cc]: What if `m_EMSOverrideCoilSpeedNumValue` is <= 0?

**Review by mjwitte — COMMENTED** [review]
Regarding the documentation, it's fine to include it with the unitarysystem actuators, but that section could use some subheadings or something to make it more readable.

**Comment by xuanluo113** [comment]
@mjwitte Hi Mike. I've fixed the recurring warning issues, and added some subsubsection headings in the doc.

File: src/EnergyPlus/UnitarySystem.cc [hunk 4]
```diff
@@ -11933,7 +11934,90 @@ namespace UnitarySystems {
...
+                        this->m_SpeedNum = this->m_NumOfSpeedCooling;
+                        CycRatio = this->m_CoolingCycRatio = 1.0;
+                        PartLoadFrac = SpeedRatio = this->m_CoolingSpeedRatio = 1.0;
+                        useMaxedSpeed = true;
+                        if (this->m_CoilSpeedErrIdx == 0) {
+                            ShowWarningMessage(state, "Wrong coil speed EMS override value, for unit=\"" + this->m_CoolingCoilName);
+                            ShowContinueError(state,
+                                              "  Exceeding maximum coil speed level. Speed level is set to the maximum coil speed level allowed.");
+                        }
+                        ShowRecurringWarningErrorAtEnd(
```
Comment (mjwitte) [inline src/EnergyPlus/UnitarySystem.cc:11953]: This is better. I was going to comment that the two error messages here will conflict once the `m_CoilSpeedErrIdx` is set. But now I see that `ShowRecurringWarningErrorAtEnd` ignores the incoming index and searches for a matching error string. A test file which sets the coil speed to values higher than max peed, and to negative values, does produce both recurring error messages. So, this is fine as-is.

File: src/EnergyPlus/UnitarySystem.cc [hunk 5]
```diff
@@ -12281,6 +12292,35 @@ namespace UnitarySystems {
                         } else if (this->m_DehumidificationMode == 1) {
                             OperationMode = DataHVACGlobals::coilEnhancedMode;
                         }
+
+                        //                        if (this->m_EMSOverrideCoilSpeedNumOn) {
```
Comment (mjwitte) [inline src/EnergyPlus/UnitarySystem.cc]: There's a block of commented code here that needs to be removed.

File: testfiles/CMakeLists.txt [hunk 6]
```diff
@@ -249,10 +249,12 @@ add_simulation_test(IDF_FILE EMSCustomOutputVariable.idf EPW_FILE USA_IL_Chicago
 add_simulation_test(IDF_FILE EMSCustomSchedule.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE EMSDemandManager_LargeOffice.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE EMSDiscreteAirSystemSizes.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
+add_simulation_test(IDF_FILE EMSLoadBasedMultiSpeedDXCoilOverrideControl.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
```
Comment (mjwitte) [inline testfiles/CMakeLists.txt:252]: This new test file has some warnings:
```
   *************  ** Warning ** SimHVAC: Exceeding Maximum iterations for all HVAC loops, during JAN14 continues
   *************  **   ~~~   **   This error occurred 44 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************
   *************  ** Warning ** Wrong coil speed EMS override value, for unit="SYS 1 FURNACE DX COOL HEATING COIL". Exceeding maximum coil speed level. Speed level is set to the maximum coil speed level allowed.
   *************  **   ~~~   **   This error occurred 63 total times;
   *************  **   ~~~   **   during Warmup 54 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************  **   ~~~   **   Max=1.  Min=1.
```

**Review by mjwitte — COMMENTED** [review]
This is better, but there is a failed unit test now.

Thanks for adding the subheadings in the unitary equipment EMS actuators section. Please add one more "Unitary System Sizing" before the paragraph starting with "Actuators are available for overriding the autosize values".

**Review by mjwitte — APPROVED** [review]
@xuanluo113 
Docs are good.
Unit tests pass.
Output variable warnings are gone (I just pushed up that change for the other idf).
Still troubled by these warnings in EMSLoadBasedMultiSpeedDXCoilOverrideControl:
```
   *************  ** Warning ** SimHVAC: Exceeding Maximum iterations for all HVAC loops, during JAN14 continues
   *************  **   ~~~   **   This error occurred 44 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************
   *************  ** Warning ** Wrong coil speed EMS override value, for unit="SYS 1 FURNACE DX COOL HEATING COIL". Exceeding maximum coil speed level. Speed level is set to the maximum coil speed level allowed.
   *************  **   ~~~   **   This error occurred 63 total times;
   *************  **   ~~~   **   during Warmup 54 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************  **   ~~~   **   Max=1.  Min=1.
   *************
   *************  ** Warning ** SimHVAC: Exceeding Maximum iterations for all HVAC loops, during JULY7 continues
   *************  **   ~~~   **   This error occurred 4 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
 ```
If there's not an obvious fix for these, please post a followup issue to investigate further.

**Comment by mjwitte** [comment]
Merging this. Will post a followup issue about the warnings in EMSLoadBasedMultiSpeedDXCoilOverrideControl.

**Pull request merged**